File: Python/backend_flask_archi_feature/author/db/author_inmemory_mapper.py
from datetime import date
from typing import Dict
from shared.mapper import BaseMapper
from author.author import Author
from author.db.author_inmemory import AuthorInMemory

class AuthorInMemoryMapper(BaseMapper[Author, AuthorInMemory]):
    def to_domain(self, external: AuthorInMemory) -> Author:
        return Author.create(
            id=external.id,
            first_name=external.first_name,
            last_name=external.last_name,
            birth_date=external.birth_date,
            nationality=external.nationality,
            death_date=external.death_date,
            biography=external.biography,
            books=external.books
        )

    def to_external(self, domain: Author) -> AuthorInMemory:        
        return AuthorInMemory(
            id=domain.id,
            first_name=domain.first_name,
            last_name=domain.last_name,
            birth_date=domain.birth_date,
            nationality=domain.nationality,
            death_date=domain.death_date,
            biography=domain.biography,
            books=domain.books
        )

File: Javascript/backend_nest/src/infra/db/inmemory/mappers/AuthorInMemoryMapper.ts
import { Author } from 'src/domain/entities/Author';
import { Mapper } from 'src/domain/kernel/Mapper';
import { AuthorInMemory } from '../entities/AuthorInMemory';

export class AuthorInMemoryMapper extends Mapper<Author, AuthorInMemory> {
  toDomain(external: AuthorInMemory): Author {
    return Author.create({
      id: external.id,
      firstName: external.firstName,
      lastName: external.lastName,
      birthDate: new Date(external.birthDate),
      nationality: external.nationality,
      deathDate: external.deathDate ? new Date(external.deathDate) : null,
      biography: external.biography,
      books: [],
    });
  }

  toExternal(domain: Author): AuthorInMemory {
    return new AuthorInMemory(
      domain.id,
      domain.firstName,
      domain.lastName,
      domain.birthDate.toISOString().split('T')[0], // Conversion Date vers string "YYYY-MM-DD"
      domain.nationality,
      domain.deathDate ? domain.deathDate.toISOString().split('T')[0] : null,
      domain.biography
    );
  }
}
